codex/purchasing/states/descriptions.nim

```diff
@@ -0,0 +1,27 @@
+import ../statemachine
+import ./cancelled
+import ./error
+import ./failed
+import ./finished
+import ./pending
+import ./started
+import ./submitted
+
+proc description*(state: PurchaseState): string =
+  if state of PurchaseCancelled:
+    "cancelled"
+  elif state of PurchaseErrored:
+    "errored"
+  elif state of PurchaseFailed:
+    "failed"
+  elif state of PurchaseFinished:
+    "finished"
+  elif state of PurchasePending:
+    "pending"
+  elif state of PurchaseStarted:
+    "started"
+  elif state of PurchaseSubmitted:
+    "submitted"
+  else:
+    "unknown"
```

Comment:
It might be a bit cleaner to create a `description` pragma which accepts a description and if none provided, defaults to the name of the of the type.
Something like this:
```nim
template describe*(name: string) {.pragma.}
    ## Specifies the description for the object

template description*(obj: auto | typedesc): string =
    when macros.hasCustomPragma(obj, describe):
        macros.getCustomPragmaVal(obj, describe)
    else:
        typetraits.name(obj.type).toLower

### Usage:

type
  PurchaseStarted* {.describe("started").} = object of PurchaseState
  PurchaseNone* = object of PurchaseState

proc test() =
  check PurchaseStarted.description == "started"
  check PurchaseNone.description == "purchasenone"
```

Reply:
Good idea. I don't think we can use custom pragmas because the actual subtype of PurchaseState is not known until runtime. I have used a `method` instead, so that we can still define the description in the same module as the type.